pkg/operator/staticpod/controller/installerstate/installer_state_controller.go

```diff
@@ -0,0 +1,218 @@
+	"k8s.io/client-go/informers"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const installerStateControllerWorkQueueKey = "key"
+
+// maxToleratedPodPendingDuration is the maximum time we tolerate installer pod in pending state
+var maxToleratedPodPendingDuration = 5 * time.Minute
+
+type InstallerStateController struct {
+	podsGetter      corev1client.PodsGetter
+	queue           workqueue.RateLimitingInterface
+	cachesToSync    []cache.InformerSynced
+	targetNamespace string
+	operatorClient  v1helpers.StaticPodOperatorClient
+	eventRecorder   events.Recorder
+
+	timeNowFn func() time.Time
+}
+
+func NewInstallerStateController(kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	podsGetter corev1client.PodsGetter,
+	operatorClient v1helpers.StaticPodOperatorClient,
+	targetNamespace string,
+	recorder events.Recorder,
+) *InstallerStateController {
+	c := &InstallerStateController{
+		podsGetter:      podsGetter,
+		targetNamespace: targetNamespace,
+		operatorClient:  operatorClient,
+		eventRecorder:   recorder.WithComponentSuffix("installer-state-controller"),
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "InstallerStateController"),
+		timeNowFn:       time.Now,
+	}
+
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Pods().Informer().HasSynced)
+	kubeInformersForTargetNamespace.Core().V1().Pods().Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+func (c *InstallerStateController) eventHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.queue.Add(installerStateControllerWorkQueueKey) },
+		UpdateFunc: func(old, new interface{}) { c.queue.Add(installerStateControllerWorkQueueKey) },
+		DeleteFunc: func(obj interface{}) { c.queue.Add(installerStateControllerWorkQueueKey) },
+	}
+}
+
+// degradedConditionNames lists all supported condition types.
+var degradedConditionNames = []string{
+	"InstallerPodPendingDegraded",
+	"InstallerPodContainerWaitingDegraded",
+}
+
+func (c *InstallerStateController) sync() error {
+	pods, err := c.podsGetter.Pods(c.targetNamespace).List(metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set{"app": "installer"}).String(),
+	})
+	if err != nil {
+		return err
+	}
+
+	// collect all pods that are in pending state for longer than maxToleratedPodPendingDuration
+	pendingPods := []*v1.Pod{}
+	hasPendingPods := false
+	for _, pod := range pods.Items {
+		if pod.Status.Phase != v1.PodPending {
+			continue
+		}
+		// we need to track this so we can requeue faster when there are pending pods
+		hasPendingPods = true
```

Comment:
we're going to want that node name for a message

Reply:
@deads2k isn't node name included in the pod name?

Reply:
nvmd. added